sycl/test/check_device_code/device_global_const_eval_use.cpp

```diff
@@ -0,0 +1,104 @@
+// RUN: %clangxx -std=c++20 -fsycl -c -fsycl-device-only -S -emit-llvm %s -o - | FileCheck %s
+
+// Tests that device_globals with device_image_scope property can be compile
+// time constant initialized and that the constant value appears in the IR
+
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+
+// Constant int and array of ints device_globals
+constexpr device_global<int, decltype(properties(device_image_scope))> 
+   dg_int{5};
+constexpr device_global<int[3], decltype(properties(device_image_scope))>
+   dg_int_arr{5, 2, 3};
+
+// Constant float and array of float device_globals
+constexpr device_global<float, decltype(properties(device_image_scope))>
+   dg_float{4.5};
+constexpr device_global<float[6], decltype(properties(device_image_scope))>
+   dg_float_arr{4.5, 2.1, 3.5, 9.33, 2.33, 2.1};
+   
+// Constant double and array of double device_globals
+constexpr device_global<double, decltype(properties(device_image_scope))>
+   dg_double{3.56543};
+constexpr device_global<double[3], decltype(properties(device_image_scope))>
+   dg_double_arr{2.2341234, 233.23423, 236.52321};
+
+// Constant bool and array of bool device_globals
+constexpr device_global<bool, decltype(properties(device_image_scope))>
+   dg_bool{true};
+constexpr device_global<bool[3], decltype(properties(device_image_scope))>
+   dg_bool_arr{true, false, true};
+
+// Constant struct and array of struct device_globals
+struct TestStruct {
+  int field1;
+  bool field2;
+  float field3;
+  int field4[4];
+};
+constexpr TestStruct TS1(5, true, 2.1, {1, 2, 3, 4});
+constexpr TestStruct TS2(7, false, 2.4, {1, 2, 3, 4});
+constexpr TestStruct TS3(6, false, 4.34534, {5, 6, 7, 8});
+constexpr device_global<TestStruct, decltype(properties(device_image_scope))>
+   dg_struct{TS3};
+constexpr device_global<TestStruct[2], decltype(properties(device_image_scope))>
+   dg_struct_arr{TS1, TS2};
+
+int main () {
+  sycl::queue Q;
+  Q.submit([&](sycl::handler& h) {
+    // Simple kernel that just copies over the values from the device_globals so
+    // that we can observe the GlobalVariables that are created to represent
+    // them in the IR
+    h.single_task([=] {
+      // Int and array of ints
+      int test_int = dg_int;
+      int test_int_arr[3];
+      for (int i = 0; i < 3; i++) {
+        test_int_arr[i] = dg_int_arr[i];
+      }
+      // CHECK: @_ZL6dg_int = internal addrspace(1) constant { i32 } { i32 5 }, align 4, !spirv.Decorations !0 #0
+      // CHECK: @_ZL10dg_int_arr = internal addrspace(1) constant { [3 x i32] } { [3 x i32] [i32 5, i32 2, i32 3] }, align 4, !spirv.Decorations !2 #1
```

Comment:
i would cleanup the highly variable data:
* the mangled names of variables - just look for the original name without the mangled prefix
* remove the `!2 #1` after the `!spirv.Decorations`